### PR TITLE
Disallow arrays of System.Void

### LINKED ISSUE #94086: [Mono] System.Runtime.Tests fail with TypeLoadException
In https://github.com/dotnet/runtime/pull/88620 I've added a very simple code that performs following checks:

```cs
private static void ValidateArrayType(Type arrayType)
{
    if (!arrayType.IsArray)
        ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_HasToBeArrayClass, ExceptionArgument.arrayType);

    ValidateElementType(arrayType.GetElementType()!);
}

private static void ValidateElementType(Type elementType)
{
    while (elementType.IsArray)
    {
        elementType = elementType.GetElementType()!;
    }

    if (elementType.IsByRef || elementType.IsByRefLike)
        ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_ByRefLikeArray);
    if (elementType == typeof(void))
        ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_VoidArray);
    if (elementType.ContainsGenericParameters)
        ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_OpenType);
}
```

The tests pass on CLR and NativeAOT, but fail on Mono:

```cs
    System.Tests.ArrayTests.CreateInstanceFromArrayType_NotSupportedArrayType_ThrowsNotSupportedException(elementType: typeof(System.Void)) [FAIL]
      Assert.Throws() Failure: Exception type was not an exact match
      Expected: typeof(System.NotSupportedException)
      Actual:   typeof(System.TypeLoadException)
      ---- System.TypeLoadException : Could not load array element class, due to: Error getting the interfaces of System.Void[] due to Error Loading class assembly:/root/helix/work/correlation/shared/Microsoft.NETCore.App/9.0.0/System.Private.CoreLib.dll type:IList`1 member:(null) assembly:/root/helix/work/correlation/shared/Microsoft.NETCore.App/9.0.0/System.Private.CoreLib.dll type:Void[] member:(null)
      Stack Trace:
        /_/src/libraries/System.Runtime/tests/System/ArrayTests.cs(1957,0): at System.Tests.ArrayTests.CreateInstanceFromArrayType_NotSupportedArrayType_ThrowsNotSupportedException(Type elementType)
        /_/src/mono/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.Mono.cs(22,0): at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args)
        /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs(178,0): at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
        ----- Inner Stack Trace -----
        /_/src/mono/System.Private.CoreLib/src/System/RuntimeTypeHandle.cs(287,0): at System.RuntimeTypeHandle.GetElementType(RuntimeType type)
        /_/src/libraries/System.Private.CoreLib/src/System/RuntimeType.cs(96,0): at System.RuntimeType.GetElementType()
        /_/src/libraries/System.Private.CoreLib/src/System/Array.cs(338,0): at System.Array.ValidateArrayType(Type arrayType)
        /_/src/libraries/System.Private.CoreLib/src/System/Array.cs(261,0): at System.Array.CreateInstanceFromArrayType(Type arrayType, Int32[] lengths)
        /_/src/libraries/System.Runtime/tests/System/ArrayTests.cs(1957,0): at System.Tests.ArrayTests.<>c__DisplayClass90_0.<CreateInstanceFromArrayType_NotSupportedArrayType_ThrowsNotSupportedException>b__1()
    System.Threading.Tests.PeriodicTimerTests.PeriodicTimer_NoActiveOperations_TimerNotRooted [SKIP]
```

What is really strange is that due to this failure, other tests that were passing so far started to fail as well:

```log
    System.Tests.TypeTests.IsSZArray_TrueForSZArrayTypes [FAIL]
      System.TypeLoadException : Error getting the interfaces of System.Void[] due to Error Loading class assembly:/root/helix/work/correlation/shared/Microsoft.NETCore.App/9.0.0/System.Private.CoreLib.dll type:IList`1 member:(null)
      Stack Trace:
        /_/src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs(1900,0): at System.RuntimeType.MakeArrayType()
        /_/src/libraries/System.Runtime/tests/System/Type/TypeTests.cs(613,0): at System.Tests.TypeTests.<>c.<IsSZArray_TrueForSZArrayTypes>b__36_0(Type nonArrayBaseType)
        /_/src/libraries/System.Linq/src/System/Linq/Select.cs(238,0): at System.Linq.Enumerable.SelectListIterator`2[[System.Type, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Type, System.Private.CoreLib, Version=9.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
        /_/src/libraries/System.Runtime/tests/System/Type/TypeTests.cs(613,0): at System.Tests.TypeTests.IsSZArray_TrueForSZArrayTypes()
        /_/src/mono/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.Mono.cs(22,0): at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args)
        /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs(57,0): at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
    System.Threading.Tests.PeriodicTimerTests.PeriodicTimer_ActiveOperations_TimerRooted [SKIP]
```

I was **not** able to repro it with a standalone app:

```cs
using System.Reflection;

namespace ArrayOfVoid
{
    internal class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine($"IsMono: {Type.GetType("Mono.RuntimeStructs") != null}");

            Console.WriteLine(typeof(void).MakeArrayType().IsArray);
            Console.WriteLine(typeof(void).MakeArrayType().MakeArrayType().IsArray);

            Console.WriteLine(typeof(void).MakeArrayType().GetElementType()!.IsArray);
            Console.WriteLine(typeof(void).MakeArrayType().MakeArrayType().GetElementType()!.IsArray);
            Console.WriteLine(typeof(void).MakeArrayType().MakeArrayType().GetElementType()!.GetElementType()!.IsArray);
        }
    }
}
```

```cmd
dotnet publish -c Release -r win-x64 --self-contained /p:UseMonoRuntime=true
.\bin\Release\net8.0\win-x64\publish\ArrayOfVoid.exe
```

```log
IsMono: True
True
True
False
True
False
```

This test is trying to create an array of arrays of void, I suspect that this is somehow triggering some edge case.

BTW some of the [tests](https://github.com/search?q=repo%3Adotnet%2Fruntime%2052072&type=code) are marked with `[ActiveIssue("https://github.com/dotnet/runtime/issues/52072"` despite the fact that https://github.com/dotnet/runtime/issues/52072 has been closed (cc @mdh1418)

### PR DESCRIPTION
**Pull request by jkotas**
The CoreCLR type loader allowed creating arrays of `System.Void`. Many operations with these invalid array types failed, often in inscrutable ways. For example, `GetInterfaces()` call failed with type load exception of `IEnumerable<void>` type. The exact failure modes are different between runtimes.

It is better to disallow creating these invalid array types in the first place, across all runtimes, to make the behavior robust and consistent.

Related to #88620

Fixes #94086

### CODE REVIEW

File: src/mono/mono/metadata/icall.c [hunk 1]
```diff
@@ -5966,9 +5966,9 @@ check_for_invalid_array_type (MonoType *type, MonoError *error)
 	gboolean allowed = TRUE;
 	char *name;
 
-	if (m_type_is_byref (type))
+	if (MONO_TYPE_IS_VOID (type))
```
Comment (lambdageek) [inline src/mono/mono/metadata/icall.c:5969]: I think this only handles reflection. For IL loading we either need a check in `mono_metadata_parse_array_internal` or around here in `mono_class_create_bounded_array`:

https://github.com/dotnet/runtime/blob/9bfd039f22e92d8ef46384945955470a8dbd80b4/src/mono/mono/metadata/class-init.c#L1173-L1184

I can't tell without running some tests which is the "right" place to change.  I think it's `mono_class_create_bounded_array`, but I'm not certain.

File: src/mono/mono/metadata/class-init.c [hunk 2]
```diff
@@ -1170,12 +1170,8 @@ mono_class_create_bounded_array (MonoClass *eclass, guint32 rank, gboolean bound
 	klass->rank = GUINT32_TO_UINT8 (rank);
 	klass->element_class = eclass;
 
-	if (m_class_get_byval_arg (eclass)->type == MONO_TYPE_TYPEDBYREF) {
```
Comment (jkotas) [inline src/mono/mono/metadata/class-init.c:1173]: TypedReference is byreflike type, so it is not necessary to check for it explicitly here.

**Comment by VSadov** [comment]
I thought this was always disallowed.

**Review by VSadov — APPROVED** [review]
LGTM !!

**Comment by VSadov** [comment]
>  For example, GetInterfaces() call failed with type load exception of `IEnumerable<void>` type.

Does that work with open generic types or pointers?

**Comment by jkotas** [comment]
> Does that work with open generic types or pointers?

Yes, it does. Pointers are special cased here: https://github.com/dotnet/runtime/blob/45978bd7191fcbb793dc12b162540022063551b6/src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs#L959

**Comment by vargaz** [comment]
The mono changes look ok to me.

**Comment by jkotas** [comment]
Breaking change issue https://github.com/dotnet/docs/issues/38311

**Pull request merged**